lib/writers.js

```diff
@@ -194,63 +202,56 @@ class WriteQueue extends events.EventEmitter {
 
   run() {
     if (!this.isRunning) {
-      this.process();
+      this.isRunning = true;
+      // Use nextTick to allow the queue to build up on the current phase
+      process.nextTick(() => this.process());
     }
   }
 
   process() {
-    const self = this;
-    utils.whilst(
-      function condition() {
-        return self.queue.length > 0;
-      },
-      function whileProcess(next) {
-        self.isRunning = true;
-        const buffers = [];
-        const callbacks = [];
-        let totalLength = 0;
-        while (totalLength < self.coalescingThreshold && self.queue.length > 0) {
-          const writeItem = self.queue.shift();
-          try {
-            const data = writeItem.operation.request.write(self.encoder, writeItem.operation.streamId);
-            totalLength += data.length;
-            buffers.push(data);
-            callbacks.push(writeItem.callback);
-          }
-          catch (err) {
-            writeItem.callback(err);
-            // Break and flush what we have
-            break;
-          }
-        }
-        if (buffers.length === 0) {
-          // No need to invoke socket.write()
-          return next();
-        }
-        // Before invoking socket.write(), mark that the request has been written to avoid race conditions.
-        for (let i = 0; i < callbacks.length; i++) {
-          callbacks[i]();
-        }
-        self.netClient.write(Buffer.concat(buffers, totalLength), function socketWriteCallback(err) {
-          if (err) {
-            self.setWriteError(err);
-          }
-          // Allow IO between writes
-          setImmediate(next);
-        });
-      },
-      function loopFinished() {
-        // The queue is now empty
-        self.isRunning = false;
+    this.netClient.cork();
+    let totalLength = 0;
+    let willEmitDrain = false;
+    while (this.queue.length > 0 && totalLength < this.coalescingThreshold && !this.error) {
+      const writeItem = this.queue.shift();
+      if (!writeItem.operation.canBeWritten()) {
+        // Invoke the write callback with an error that is not going to be yielded to user
+        // as the operation has timed out or was cancelled.
+        writeItem.callback(new Error('The operation was already cancelled or timeout elapsed'));
+        continue;
+      }
+      let data;
+      try {
+        data = writeItem.operation.request.write(this.encoder, writeItem.operation.streamId);
+      }
+      catch (err) {
+        writeItem.callback(err);
+        continue;
       }
-    );
+      totalLength += data.length;
+      // Mark it as written to avoid race conditions
+      writeItem.callback();
+      willEmitDrain = !this.netClient.write(data, err => this.setWriteError(err));
+    }
+    // Flush all data
+    this.netClient.uncork();
+    if (!willEmitDrain) {
```

Comment:
For my benefit, the idea here is that if successive calls to `this.netClient.write` all return true, we won't get a drain event because the stream is not getting backed up.  In this case we call `run` immediately to continue consuming the write queue.  Otherwise, we wait for the `drain` event (line 176) which is our signal to know that we can continue consuming the write queue and writing.

Reply:
I've simplified the logic a little bit and renamed the variable to `fitsInStreamBuffer` to avoid negating it.

Check it out and let me know what you think.

The idea is that the queue is marked as "running" when its actually writing (sync) and when its waiting for the `drain` event.

When there are enough items in the queue, the flow is:
`process() -> 'drain' event -> process() -> 'drain' event -> ...`